chore: refactor token registry

rjchow opened the pull request:
- moved token registry state into TokenRegistry class
- moved Token state into Token
- added tests for the smart contract parts of token registry

Nebulis reviewed (approved):
Please update the README

Nebulis commented on src/registry.ts:
```diff
@@ -0,0 +1,41 @@
+
+  address: string;
+
+  contractInstance: any;
```
type Contract from ethers ?

rjchow replied:
ok

Nebulis commented on src/registry.ts:
```diff
@@ -0,0 +1,41 @@
+
+  /**
+   * Creates a TokenRegistry instance with the specified address and ethersjs signer
+   * @param {object} instanceParameters
```
you shouldnt type the parameter in jsdoc (easy to make the type non synchronized)

rjchow replied:
ok

Nebulis commented on src/registry.ts:
```diff
@@ -0,0 +1,41 @@
+    this.contractInstance = new ethers.Contract(contractAddress, JSON.stringify(TokenRegistryABI), web3Provider);
+  }
+
+  async mint(document: Document, owner: string) {
```
- do you mind trying to upgrade to open-attestation v3
- isn't it a wrapped document ? (or should it ?)

rjchow replied:
must I upgrade? can I wait until you sort out v3? :D
it is a wrapped document indeed

Nebulis replied:
it might be worth to still upgrade to open-attestation v3 as the typing has been improved, it doesnt mean you need to handle all document you can only focus on v2 document: `WrappedDocument<v2.OpenAttestationDocument>`. We can eventually go through this together

Dont forget to use getData if you handle WrappedDocument

rjchow replied:
i don't need to use getData because i'm using the content of the wrapper

Nebulis replied:
:ok:

Nebulis commented on src/registry.ts:
```diff
@@ -0,0 +1,41 @@
+  async transferTo(document: Document, newOwner: string) {
+    const tokenId = getBatchMerkleRoot(document);
+    const currentOwner = await this.ownerOf(document);
+    return this.contractInstance["safeTransferFrom(address,address,uint256)"](currentOwner, newOwner, tokenId);
```
does it look like :hankey: because there are multiple `safeTransferFrom` with different signature ?

rjchow replied:
exactly

Nebulis commented on src/token.ts:
```diff
@@ -0,0 +1,29 @@
+export class Token {
+  document: Document;
+
+  web3Provider: any;
```
type Signer?

Nebulis commented on src/token.ts:
```diff
@@ -0,0 +1,29 @@
+  constructor(document: Document, web3Provider?: any) {
+    this.document = document;
+    this.web3Provider = web3Provider;
+    this.tokenRegistry = new TokenRegistry({ contractAddress: getIssuer(document).tokenRegistry, web3Provider });
```
`web3Provider` is not an optional parameter in `TokenRegistry` but it's here, is it expected ?

Nebulis replied:
It looks expected according to the TODO below, but I'm surprised TS doesnt complain :grimacing:

Nebulis commented on src/util/token.ts:
```diff
@@ -1,23 +1,15 @@
-  const contractInstances = getContractInstance(issuer);
-  return contractInstances.ownerOf(tokenId);
+export const getBatchMerkleRoot = (document: Document) => {
+  return `0x${get(document, "signature.merkleRoot")}`;
```
document.signature.merkleRoot ?

Nebulis commented on src/util/token.test.ts:
```diff
@@ -0,0 +1,35 @@
+      expect(issuer).toEqual(expectedIssuer);
+    });
+
+    it("should throw an error if there is more than 1 issuer", () => {
```
why does the schema allow multiple issuers if it's forbidden ?

rjchow replied:
we can forbid multiple issuers on token type documents no issue

Nebulis commented on test/registry.js:
```diff
@@ -0,0 +1,48 @@
+    const newOwner = await signer2.getAddress();
+    await tokenRegistryInstance.mint(documentStub, newOwner);
+    const currentOwner = await tokenRegistryInstance.ownerOf(documentStub);
+    expect(currentOwner).to.eql(newOwner);
```
to.deep.equal ? =) (everywhere)

rjchow replied:
AAAAAAAAAAAAAAAAAAAAHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHHH

Nebulis commented on test/registry.js:
```diff
@@ -0,0 +1,48 @@
+  beforeEach("", async () => {
+    ERC721Instance = await ERC721.new("foo", "bar");
+    ERC721Address = ERC721Instance.address;
+    provider = new ethers.providers.Web3Provider(web3.currentProvider);
```
PP I dont understand this =)

Nebulis commented on test/registry.js:
```diff
@@ -0,0 +1,48 @@
+    const nextOwner = await tokenRegistryInstance2.ownerOf(documentStub);
+    expect(nextOwner).to.eql(nextOwnerAddress);
+  });
+});
```
it should get owner of ?

rjchow committed:
chore: refactored oa-token entirely

created classes ReadOnlyToken and WriteableToken for separating concerns between read operations and write operations
created tests against ganache for these classes
tidied up a whole bunch of stuff

john-dot-oa commented:
:tada: This PR is included in version 1.2.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@govtechsg/oa-token/v/1.2.0)
- [GitHub release](https://github.com/Open-Attestation/oa-token/releases/tag/v1.2.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
